mypy/checkmember.py

```diff
@@ -1203,9 +1203,15 @@ def analyze_class_attribute_access(
         t = get_proper_type(t)
         if isinstance(t, FunctionLike) and is_classmethod:
             t = check_self_arg(t, mx.self_type, False, mx.context, name, mx.msg)
-        result = add_class_tvars(
+        t = add_class_tvars(
             t, isuper, is_classmethod, is_staticmethod, mx.self_type, original_vars=original_vars
         )
+        if is_decorated and not is_staticmethod:
+            t = expand_self_type_if_needed(
```

Comment:
I don't quite understand why we need to put this self type expansion here. Specifically: this branch is only for decorators. Shouldn't whatever is applying the decorators handle self types instead?

Reply:
This function handles lookup on classes, we bind self types here for everything that can reasonably contain them as this is the latest moment possible (see `Var` branch above). Consider a decorator that preserves function signature applied to a `def method(self) -> Self`. Applying the decorator still yields a `Self`-returning callable, which is the best representation until we actually want to show it or apply it to something, when `Self` loses its semantic meaning. Otherwise, for example, that method will be impossible to cleanly override in subclasses.

Reply:
Alright that makes sense then. I didn't realize when this function was called.